### PR TITLE
Fix: prevent duplicate istioEndpoints generated

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

After #51279

### CODE REVIEW

**Review by leosarra — APPROVED** [review]
LGTM, ty for fixing it

**Comment by zirain** [comment]
there's no test for this?

**Comment by hzxuzhonghu** [comment]
Not any ut. Opened an issue to track found https://github.com/istio/istio/issues/52107

**Comment by leosarra** [comment]
If you want I can look into the unit tests for buildIstioEndpointsWithService and updateEndpointCacheForSlice next week (this week I don't have much availability)

**Comment by hzxuzhonghu** [comment]
Thank you @leosarra.

File: pilot/pkg/serviceregistry/kube/controller/endpointslice.go [hunk 1]
```diff
@@ -244,29 +245,45 @@ func endpointHealthStatus(svc *model.Service, e v1.Endpoint) model.HealthStatus
...
+	if epSlice.AddressType == v1.AddressTypeFQDN {
 		// TODO(https://github.com/istio/istio/issues/34995) support FQDN endpointslice
 		return
 	}
 	svc := esc.c.GetService(hostName)
-	svcNamespacedName := getServiceNamespacedName(slice)
+	svcNamespacedName := getServiceNamespacedName(epSlice)
 	svcCore := esc.c.services.Get(svcNamespacedName.Name, svcNamespacedName.Namespace)
+	if svcCore != nil && len(svcCore.Spec.ClusterIPs) > 1 {
+		// It means this is dual stack service, for which k8s will build two endpointslice per pod.
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/endpointslice.go:258]: I think its possible to have a dualstack service which only has 1 IPv6 address. I doubt kubernetes endpoint controller will generate it, but users can create endpoints by hand and we need to support those.

Comment (leosarra) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:258]: Ah good catch, users could make a service without a selector and then manually create only an IPv6 EndpointSlices

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:258]: Help me understand, how could dual stack service have one clusterIP?

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:258]: I do not mean 1 ClusterIp, but only 1 endpointslice with only ipv6

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:258]: Per k8s, i think if the service is ipv6, it will by default generate iptables with only ipv6. But for ipv4, it generated iptables for ipv4. 

For the case, user created endpointslice themself,  i think they must obet the service ip family. I feel if a service with ipv6 family, but user create endpointslice with ipv4 address, they could not access.

So what is missing for this case, i am not quite sure

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:258]: I don't think there is anything suggesting a dual stack family has to have a pair of v4-v6 endpoints for everything it's representing. There is no correlation between the two - they are two entirely orthogonal sets of endpoints. 

Istio is artificially joining the endpoints with pods based on a heuristic (for general use, not just dual stack), and now for dual stack we are *joining* these based on the pod. 

But this PR does not do a full join, it only uses the v4 on. So if there is no v4 at all, which is valid, we miss it. 

These slices may not even be bound to pods at all, in which case this is even more broke since there is no "join" happening. In that case we just lose all of the Ipv6 endpoints entirely since we couldn't recover them. 

Obviously we have a huge testing gap for these edge cases.. happy to help expand those

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:258]: TBH the pod case is probably not going to happen in real world, but the non-pod case is just dropping all the Ipv6 endpoints entirely here which is not even an edge case

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:258]: Not sure that made sense, I can provide a concrete test case when I'm at a computer if it will help

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:258]: I expect a concrete example to be more clear

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:258]: BTW, the below check does not mean, we only generate we ignore ipv6, becasue we acquire the ips for pod at last. So all ips in pod status will be there. 

```
if epSlice.AddressType == v1.AddressTypeIPv6 {
			return
		}
```

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:258]: I am mostly worried about endpoints that are not for Pods at all.

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:258]: IC, indeed missed processing

File: pilot/pkg/serviceregistry/kube/controller/endpointslice.go [hunk 2]
```diff
@@ -244,29 +245,55 @@ func endpointHealthStatus(svc *model.Service, e v1.Endpoint) model.HealthStatus
...
 	var endpoints []*model.IstioEndpoint
-	if slice.AddressType == v1.AddressTypeFQDN {
+	if epSlice.AddressType == v1.AddressTypeFQDN {
 		// TODO(https://github.com/istio/istio/issues/34995) support FQDN endpointslice
 		return
 	}
 	svc := esc.c.GetService(hostName)
-	svcNamespacedName := getServiceNamespacedName(slice)
+	svcNamespacedName := getServiceNamespacedName(epSlice)
+	// This is not a endpointslice for service, ignore
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/endpointslice.go:256]: what if it used to have the label? Is this fixing a bug?

File: pilot/pkg/serviceregistry/kube/controller/endpointslice.go [hunk 3]
```diff
@@ -244,29 +245,55 @@ func endpointHealthStatus(svc *model.Service, e v1.Endpoint) model.HealthStatus
...
+	if svcNamespacedName.Name == "" {
+		return
+	}
 	svcCore := esc.c.services.Get(svcNamespacedName.Name, svcNamespacedName.Namespace)
+	if svcCore != nil && len(svcCore.Spec.ClusterIPs) > 1 {
+		// It means this is dual stack service, for which k8s will build two endpointslice per pod.
+		// If the service is with selector, k8s will manually generate two ip families endpointslices,
+		// we ignore ipv6 family address to prevent generating duplicate IstioEndpoints.
+		if svcCore.Spec.Selector != nil {
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/endpointslice.go:265]: This is not a safe assumption. There is no special logic for service having a selector, they can still have extra endpoints added.

Full example, which works in k8s:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: shell
spec:
  selector:
    matchLabels:
      app: shell
  template:
    metadata:
      labels:
        app: shell
    spec:
      terminationGracePeriodSeconds: 0
      containers:
      - name: shell
        image: howardjohn/shell
        imagePullPolicy: IfNotPresent
        args:
          - /bin/sleep
          - infinity
        resources:
          requests:
            cpu: 10m
            memory: 128Mi
          limits:
            cpu: 500m
            memory: 500M
---
apiVersion: v1
kind: Service
metadata:
  name: shell
  labels:
    app: shell
spec:
  ports:
  - port: 9087
    name: http
  selector:
    app: shell
---
addressType: IPv4
apiVersion: discovery.k8s.io/v1
endpoints:
- addresses:
  - 1.2.3.4
  conditions:
    ready: true
    serving: true
    terminating: false
kind: EndpointSlice
metadata:
  labels:
    app: shell
    kubernetes.io/service-name: shell
  name: shell-manual
  namespace: default
ports:
- name: http
  port: 9087
  protocol: TCP
```

**Review by howardjohn — CHANGES_REQUESTED** [review]
This is looking closer but I think still has some edge cases to handle. We really need thorough test coverage here

File: pilot/pkg/serviceregistry/kube/controller/endpointslice.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:256]: We cannot remove the label, k8s will add it back

Yes, i am fixing to ignore no svc endpointslice.

File: pilot/pkg/serviceregistry/kube/controller/endpointslice.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:265]: For a service without selector, we  cannot do kind of dedup gracefully.  So ignore the rare case here.

Actuallt a urgly way to do dedup is to iterate over all the endpointslice, and group all the endpoints which belong to a same service. Then for each group, find out whether dual stack endpoints with pod are there.

I think we donot need to process this case, because i think with real pods there, normally people cannot manage the service and endpointslice manually themselves

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:265]: I don't think it's acceptable to incorrectly process endpoints, even if they are "weird" cases. In ambient we initially made assumptions like this and found in real world they ended up breaking real applications. 

It is critical to be correct here or we will cause an outage - and this impacts all users, not just ones with the experiment dual stack mode

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:265]: Do you agree with the above processing or have other suggestion


The above process could have very bad performance

**Comment by howardjohn** [comment]
I'll think about some other solutions

**Comment by hzxuzhonghu** [comment]
@howardjohn Do you have time to think about a better solution?

File: pilot/pkg/serviceregistry/kube/controller/endpointslice.go (hunk 3 above)
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:265]: I don't think we should have logic around `selector`, just about targetRef. Otherwise this lgtm. Maybe we can fix in a followup?

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/serviceregistry/kube/controller/endpointslice.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/endpointslice.go:265]: Yes, i kind of agree, with the tests  i can do more investigate
